File: ce/snippets/csharp/CRMV8/authenticatewithnohelp/cs/authenticatewithnohelp4.cs
// <snippetauthenticatewithnohelp4>


 private TProxy GetProxy<TService, TProxy>(
     IServiceManagement<TService> serviceManagement,
     AuthenticationCredentials authCredentials)
     where TService : class
     where TProxy : ServiceProxy<TService>
 {
     Type classType = typeof(TProxy);

     if (serviceManagement.AuthenticationType !=
         AuthenticationProviderType.ActiveDirectory)
     {
         AuthenticationCredentials tokenCredentials =
             serviceManagement.Authenticate(authCredentials);
         // Obtain discovery/organization service proxy for Federated, LiveId and OnlineFederated environments. 
         // Instantiate a new class of type using the 2 parameter constructor of type IServiceManagement and SecurityTokenResponse.
         return (TProxy)classType
             .GetConstructor(new Type[] { typeof(IServiceManagement<TService>), typeof(SecurityTokenResponse) })
             .Invoke(new object[] { serviceManagement, tokenCredentials.SecurityTokenResponse });
     }

     // Obtain discovery/organization service proxy for ActiveDirectory environment.
     // Instantiate a new class of type using the 2 parameter constructor of type IServiceManagement and ClientCredentials.
     return (TProxy)classType
         .GetConstructor(new Type[] { typeof(IServiceManagement<TService>), typeof(ClientCredentials) })
         .Invoke(new object[] { serviceManagement, authCredentials.ClientCredentials });
 }

// </snippetauthenticatewithnohelp4>

File: ce/customerengagement/on-premises/snippets/visualbasic/CRMV8/authenticatewithnohelpvb/vb/authenticatewithnohelp4.vb
' <snippetauthenticatewithnohelp4>


 Private Function GetProxy(Of TService As Class,
                               TProxy As ServiceProxy(Of TService)) _
                           (ByVal serviceManagement As IServiceManagement(Of TService),
                            ByVal authCredentials As AuthenticationCredentials) As TProxy
     Dim classType As Type = GetType(TProxy)

     If serviceManagement.AuthenticationType <>
         AuthenticationProviderType.ActiveDirectory Then
         Dim tokenCredentials As AuthenticationCredentials =
             serviceManagement.Authenticate(authCredentials)
         ' Obtain discovery/organization service proxy for Federated, LiveId and OnlineFederated environments. 
         ' Instantiate a new class of type using the 2 parameter constructor of type IServiceManagement and SecurityTokenResponse.
         Return CType(classType _
         .GetConstructor(New Type() {GetType(IServiceManagement(Of TService)), GetType(SecurityTokenResponse)}) _
         .Invoke(New Object() {serviceManagement, tokenCredentials.SecurityTokenResponse}), TProxy)
     End If

     ' Obtain discovery/organization service proxy for ActiveDirectory environment.
     ' Instantiate a new class of type using the 2 parameter constructor of type IServiceManagement and ClientCredentials.
     Return CType(classType _
         .GetConstructor(New Type() {GetType(IServiceManagement(Of TService)), GetType(ClientCredentials)}) _
         .Invoke(New Object() {serviceManagement, authCredentials.ClientCredentials}), TProxy)
 End Function

' </snippetauthenticatewithnohelp4>